source/iterator/iterator.go

```diff
@@ -0,0 +1,311 @@
+// Copyright © 2022 Meroxa, Inc. & Yalantis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-redshift/columntypes"
+	"github.com/conduitio-labs/conduit-connector-redshift/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/huandu/go-sqlbuilder"
+	"github.com/jmoiron/sqlx"
+	"go.uber.org/multierr"
+)
+
+const (
+	// metadataFieldTable is a name of a record metadata field that stores a Redshift table name.
+	metadataFieldTable = "redshift.table"
+	// keySearchPath is a key of get parameter of a datatable's schema name.
+	keySearchPath = "search_path"
+)
+
+// Iterator is an implementation of an iterator for Amazon Redshift.
+type Iterator struct {
+	db       *sqlx.DB
+	rows     *sqlx.Rows
+	position *Position
+
+	// table is a table name.
+	table string
+	// keyColumns is a name of the column that iterator will use for setting key in record.
+	keyColumns []string
+	// orderingColumn is a name of the column that iterator will use for sorting data.
+	orderingColumn string
+	// batchSize is a size of rows batch.
+	batchSize int
+	// columnTypes is a map with all column types.
+	columnTypes map[string]string
+}
+
+// New creates a new instance of the iterator.
+func New(ctx context.Context, driverName string, pos *Position, config config.Source) (*Iterator, error) {
+	var err error
+
+	iterator := &Iterator{
+		position:       pos,
+		table:          config.Table,
+		keyColumns:     config.KeyColumns,
+		orderingColumn: config.OrderingColumn,
+		batchSize:      config.BatchSize,
+	}
+
+	iterator.db, err = sqlx.Open(driverName, config.DSN)
+	if err != nil {
+		return nil, fmt.Errorf("open db connection: %w", err)
+	}
+
+	if iterator.position.LastProcessedValue == nil {
+		latestSnapshotValue, latestValErr := iterator.latestSnapshotValue(ctx)
+		if latestValErr != nil {
+			return nil, fmt.Errorf("get latest snapshot value: %w", latestValErr)
+		}
+
+		if config.Snapshot {
+			iterator.position.LatestSnapshotValue = latestSnapshotValue
+		} else {
+			iterator.position.LastProcessedValue = latestSnapshotValue
```

Comment:
Could you please add  small comment to explain why we do this

Reply:
I've added comments.